src/classes/OPP_OpportunityContactRoles_TDTM.cls

```diff
@@ -36,144 +36,199 @@
+                    ));
+                }
+
                 oppsInsertCR.remove(cr.OpportunityId);
             }
         }
 
         //Create CRs.
-        if (oppsInsertCR.size() > 0){
-            insertCR(oppsInsertCR.values(), dmlWrapper);
+        if (!oppsInsertCR.isEmpty()) {
+            insertCR(oppsInsertCR.values());
         }
 
         //Update existing CRs.
-        if (CRUpdates.size() > 0){
-            updateCR(CRUpdates, dmlWrapper);
+        if (!CRUpdates.isEmpty()) {
+            dmlWrapper.objectsToUpdate.addAll((List<SObject>)CRUpdates);
+        }
+
+        //fix opportunity lookup if out of sync with contact role
+        if (!fixedOpps.isEmpty()) {
+            dmlWrapper.objectsToUpdate.addAll((List<SObject>)fixedOpps);
         }
     }
 
-    /// <name> getDefaultRole </name>
-    /// <summary> Return the default role from the custom setting.  Use custom setting for default role value.</summary>
-    private static String getDefaultRole(){
+    /*******************************************************************************************************
+    * @description Return the default role from the custom setting.  Use custom setting for default role value.
+    ********************************************************************************************************/
+    private static String getDefaultRole() {
         return UTIL_CustomSettingsFacade.getContactsSettings().npe01__Opportunity_Contact_Role_Default_role__c;
     }
 
-    /// <name> insertCR </name>
-    /// <summary> Creates a new Contact Role record when an opp is inserted and there is no primary CR.  Use custom setting for default role value.</summary>
-    /// <param name="opportunties"> List of opportunities meeting trigger criteria </param>
-    private static void insertCR(Opportunity[] opportunities, DmlWrapper dmlWrapper){
+    /*******************************************************************************************************
+    * @description Creates a new Contact Role record when an opp is inserted and there is no primary CR. 
+    * Uses a custom setting for default role value.
+    * @param opportunties List of opportunities meeting trigger criteria
+    ********************************************************************************************************/
+    private void insertCR(Opportunity[] opportunities) {
 
         List<OpportunityContactRole> CRs = new List<OpportunityContactRole>();
 
         //Straightforward creation of opp contact role.      
-        for(Opportunity o : opportunities){
-            if(o.npe01__Contact_ID_for_Role__c != null) {
-                try {
-                    CRs.add(new OpportunityContactRole(OpportunityId = o.Id, ContactId = o.npe01__Contact_ID_for_Role__c, Role = getDefaultRole(), IsPrimary = true));
-                } catch (exception e) {
-                    o.addError(Label.npe01.Opportunity_Contact_Role_Error_Bad_Contact_Id);
```

Comment:
So this is where we would have generated an error about invalid contact id's.  are we now preventing this from happening?  ie, progagating an error to the integration (or more likely person importing data) that they set a bogus id.